In [ ]:
print('welcome to Lab2_Ex3_Solution!!')

welcome to Lab2_Ex3_Solution!!


In [ ]:
!pip install -q pyomo

Let u1, u2, . . . ,u10 denotes the amount(in tons) utilized from the availabe stock respectively for Alloys A1, A2, ... ,A10 and u11, u12, . . . ,u20 denotes the amount purchased(in tons) respectively for Alloys A1, A2, ... ,A10. 

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
# create a model
model = ConcreteModel()

In [ ]:
obj_coef_c = np.array([35,50,58,60,44,39,45,55,35,40,72,95,110,125,88,74,95,115,60,84])

In [ ]:
col_indices = np.arange(20)
print(col_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


In [ ]:
row_indices = np.arange(4)
print(row_indices)

[0 1 2 3]


In [ ]:
constr_coef_A = np.array([[0.80,0.75,0.75,0.60,0.55,0.55,0.40,0.35,0.30,0.30,0.80,0.75,0.75,0.60,0.55,0.55,0.40,0.35,0.30,0.30],
                          [0.15,0.15,0.10,0.20,0.25,0.10,0.50,0.15,0.30,0.55,0.15,0.15,0.10,0.20,0.25,0.10,0.50,0.15,0.30,0.55],
                          [0.05,0.10,0.15,0.20,0.20,0.35,0.10,0.50,0.40,0.15,0.05,0.10,0.15,0.20,0.20,0.35,0.10,0.50,0.40,0.15],
                          [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]])

#declaring RHS of the constraints as variables
b = np.array([112.5,87.5,50,250])

In [ ]:
#setting the lower bound and upper bounds for the variables
lower_bound = np.array([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
upper_bound = np.array([10,7,8,15,8,15,12,10,0,9,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf])

In [ ]:
#declaration of the decision variables in the model.
model.u = Var(col_indices)

In [ ]:
# creation a ConstraintList to hold multiple constraints
model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model.constraints.add(sum(constr_coef_A[i,j]*model.u[j] for j in col_indices) >= b[i])

In [ ]:
for j in col_indices:
  model.u[j].setlb(lower_bound[j])
  model.u[j].setub(upper_bound[j])

In [ ]:
#adding the model objective
model.objective = Objective(expr = sum(obj_coef_c[j]*model.u[j] for j in col_indices), sense=minimize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    u_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   20 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19}

1 Var Declarations
    u : Size=20, Index=u_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  10.0 : False :  True :  Reals
          1 :     0 :  None :   7.0 : False :  True :  Reals
          2 :     0 :  None :   8.0 : False :  True :  Reals
          3 :     0 :  None :  15.0 : False :  True :  Reals
          4 :     0 :  None :   8.0 : False :  True :  Reals
          5 :     0 :  None :  15.0 : False :  True :  Reals
          6 :     0 :  None :  12.0 : False :  True :  Reals
          7 :     0 :  None :  10.0 : False :  True :  Reals
          8 :  

In [ ]:
!apt-get install -y -qq glpk-utils

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 16238.44
  Upper bound: 16238.44
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 21
  Number of nonzeros: 81
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.019652843475341797
# ----------------------------------------------------------
#   Solution Information
# ----------------------------

In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('u[',i,'] = ', model.u[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  16238.44

Decision Variables
u[ 0 ] =  10.0
u[ 1 ] =  7.0
u[ 2 ] =  8.0
u[ 3 ] =  15.0
u[ 4 ] =  8.0
u[ 5 ] =  15.0
u[ 6 ] =  12.0
u[ 7 ] =  0.0
u[ 8 ] =  0.0
u[ 9 ] =  9.0
u[ 10 ] =  28.6
u[ 11 ] =  0.0
u[ 12 ] =  0.0
u[ 13 ] =  0.0
u[ 14 ] =  0.0
u[ 15 ] =  0.0
u[ 16 ] =  0.0
u[ 17 ] =  0.0
u[ 18 ] =  52.64
u[ 19 ] =  84.76

Constraints
constraints : Size=4
    Key : Lower : Body  : Upper
      1 : 112.5 : 112.5 :  None
      2 :  87.5 :  87.5 :  None
      3 :  50.0 :  50.0 :  None
      4 : 250.0 : 250.0 :  None


EXERCISE_3(3):The optimal solution and the value of objective function are: 
u1 =  10.0 ,
u2 =  7.0 ,
u3 =  8.0 ,
u4 =  15.0 ,
u5 =  8.0 ,
u6 =  15.0 ,
u7 =  12.0 ,
u8 =  0.0 ,
u9 =  0.0 ,
u10 =  9.0 ,
u11 =  28.6 , 
u12 =  0.0 ,
u13 =  0.0 ,
u14 =  0.0 ,
u15 =  0.0 ,
u16 =  0.0 ,
u17 =  0.0 ,
u18 =  0.0 ,
u19 =  52.64 ,
u20 =  84.76 ;
Objective =  16238.44


EXERCISE_3(4):Three alloys ranked high(decreasing order) in terms of the quantities to be bought are A-10, A-9, A-1 and the alloys ranked low are A-2, A3,A-4

EXERCISE_3(5):Alloy whose stock is completely utilized are:A1,A2,A3,A4,A5,A6,A7,A10  its obvious from the corresponding values of the variables u1,u2,u3,u4,u5,u6,u7,u10 which shows that amount available in the stock is completely utilized.

EXERCISE_3(6):alloy which need not be acquired from market are A2,A3,A4,A5,A6,A7,A8 because the corresponding variables u12,...,u18 representing the amount to be purchased from the market has value zero(0).